### PR TITLE
[add-authorization] add authorization, restructure middleware

### PR DESCRIPTION
**Pull request by pduykhang**
Hello everyone 
This pull request, I add authorization middleware, and re restructure authenticate middleware.
pls give some comments and merge this pull request

### CODE REVIEW

File: internal/pkg/middleware/authororization.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+)
+
+func Authorization(roles ...int) func(http.HandlerFunc) http.HandlerFunc {
+	l := glog.New().WithField("package", "middleware")
+	if len(roles) == 0 {
+		l.Errorf("role list must not blank")
+		return nil
+	}
+	sort.SliceIsSorted(roles, func(i, j int) bool {
+		return i > j
```
Comment (lnquy) [inline internal/pkg/middleware/authororization.go:26]: [Major] Invalid sort func. You must compare `roles[i]` with `roles[j]` instead.

File: internal/pkg/middleware/authororization.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package middleware
```
Comment (lnquy) [inline internal/pkg/middleware/authororization.go:1]: File name.

File: internal/pkg/middleware/authororization.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	jwtToken := strings.Split(theBearerToken, " ")[1]
+	JWTGeneration := jwt.NewTokenGeneration()
+	notExpire := JWTGeneration.CheckExp(jwtToken)
+	if !notExpire || !JWTGeneration.CheckValib(jwtToken) {
+		return false
+	}
+	payload := JWTGeneration.GetPayload(jwtToken)
+	jwtRolePriority := payload.Priority
+	l := glog.New().WithField("package", "middleware")
+	l.Infof("jwt priority geted : %v", jwtRolePriority)
```
Comment (lnquy) [inline internal/pkg/middleware/authororization.go:55]: Why recheck the JWT authentication here when we already have `middleware.Authentication`.
Also, this func should check the `role` for authorization only.

File: internal/pkg/middleware/authororization.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+			if ok := RoleCheck(r, roles...); !ok {
+				l.Infof("authorization fail")
+				respond.JSON(w, http.StatusUnauthorized, map[string]string{"status": "401", "message": "authorization"})
+				return
+			}
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+func RoleCheck(r *http.Request, roles ...int) bool {
```
Comment (lnquy) [inline internal/pkg/middleware/authororization.go:40]: We don't have `Moderator` role for now, but let assume that `NormalUser < Moderator < Admin`.
What if I want to allow `Admin` and `NormalUser` to access this API, but not `Moderator`, so in that case, does this func still run correctly?

Comment (pduykhang) [reply internal/pkg/middleware/authororization.go:40]: No ! When I put a additional role, which is `Moderator` this func is not correct, so I think authorization must define in detail who can do what
Example :
`middleware.Authorization(roleUser)` just user have permission, `admin` and `moderator` haven't permission
`middleware.Authorization(roleAdmin)` just admin have permission 
if you

**Review by pthethanh — APPROVED** [review]

**Pull request merged**